lib/ood_core/job/adapters/slurm.rb

```diff
@@ -483,6 +499,13 @@ def cluster_info
           @slurm.get_cluster_info
         end
 
+        # Retrieve the accounts available to use  for the current user.
+        #
+        # @return [Array<String>] the accounts available to the user.
+        def accounts
+          @slurm.accounts
+        end
```

Comment:
Why is there a `private and `public` version of this method?

Reply:
I can't tell the scope from here, but there's the outer class `Slurm < JobAdapter` and an inner class. What you've highlighted is the outer most public class. But then there's the inner object we call `@slurm` which does all the actual work.

You can see the `Batch` class (the type of object that `@slurm` is) also defined here.

Reply:
Oh, yeah, this file is hideous. Yeah sorry, i saw the private and just thought it was in that scope....but that ends, then we have an `initialize` then this.....ok, I really can't understand this design tbh so I need to talk through this.

The class instantiates itself then uses that object to call its methods on itself to build this? Sorry, I am a bit lost in this adapter.

Reply:
No it's just a nested class. I.e., a class defined within the scope of another.

Here's a simple example where you'd interact with `Outer` objects without caring about the `Inner` or any `Inner` instantiations. 
```ruby
class Outer
  def initialize
    @inner = Inner.new()
  end

  def do_something
    @inner.do_something
  end

  class Inner
    def do_something
      puts 'here I am'
    end
  end
end

Outer.new.do_something
```

Why not just put all the logic in `Outer`? The pattern was started before my time, but it's common job adapter pattern where we have another `Batch` class that actually does all the work.

It's just with the other adapters, the `Batch` classes are in other files.  They have the same pattern, here's LSF as an example where it's using an inner object called `batch`.

https://github.com/OSC/ood_core/blob/dc7bfbdc622f851194e7c7b56dadb3a61f466c95/lib/ood_core/job/adapters/lsf.rb#L106-L110

Reply:
Ah, ok, I've seen that pattern, I'm just lost in indentation in this file. Ok, that makes sense and gets this cleared up.